construct_deep_url()

RobW3LGA opened the pull request:
##### SUMMARY
- Adds the ability to create deeper objects to an ACI fabric and can potentially replace the need for the aci_rest module

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
- Modifies aci.py with construct_deep_url() and supporting functions

##### ADDITIONAL INFORMATION
- This function can also replace shallow ACI objects with a slight module change 

- Example:
```
aci.construct_url(
        root_class=dict(
            aci_class='fvTenant',
            aci_rn='tn-{0}'.format(tenant),
            module_object=tenant,
            target_filter={'name': tenant},
        ),
        subclass_1=dict(
            aci_class='fvBD',
            aci_rn='BD-{0}'.format(bd),
            module_object=bd,
            target_filter={'name': bd},
        ),
        child_classes=['fvRsCtx', 'fvRsIgmpsn', 'fvRsBDToNdP', 'fvRsBdToEpRet'],
    )
```

- vs:
```
aci.construct_deep_url(
        parent_objects=[
            dict(
                parent_class='uni',
                aci_class='fvTenant',
                aci_rn='tn-{0}'.format(tenant),
                target_filter={'name': tenant},
                module_object=tenant
            ),
        ],
        target_object=dict(
            parent_class='fvTenant',
            aci_class='fvBD',
            aci_rn='BD-{0}'.format(bd),
            target_filter={'name': bd},
            module_object=bd
        ),
        child_classes=['fvRsCtx', 'fvRsIgmpsn', 'fvRsBDToNdP', 'fvRsBdToEpRet']
    )
```

RobW3LGA committed:
Updates aci.py with the ability to add ACI objects to any depth

Changes start at line 411 (construct_deep_url() and supporting functions). One minor change to line 633 (the original construct_url()) to provide for testability:  ...join(sorted(self.child_classes)) vs ...join(self.child_classes)

I am also attaching two test files. One characterizing the existing construct_url() and the matching test set for construct_deep_url() to support my efforts and proof of parity

RobW3LGA committed:
Two PyTest files to support construct_deep_url

These two files provide testing parity, one characterizing the original construct_url() function and the other proofing construct_deep_url(). The ...deep_url.py test file goes five layers deep to provide better validation for the function

mattclay commented on lib/ansible/module_utils/network/aci/aci_deep_url_test.py:
```diff
@@ -0,0 +1,2978 @@
+import pytest
```
Test code should reside under `test/units/` and not `lib/`.

RobW3LGA replied:
Corrected and thanks for the help

RobW3LGA committed:
Correcting previous upload to incorrect folder

These two files provide testing parity, one characterizing the original construct_url() function and the other proofing construct_deep_url(). The ...deep_url.py test file goes five layers deep to provide better validation for the function

mattclay commented on lib/ansible/module_utils/network/aci/aci_construct_url_test.py:
```diff
@@ -0,0 +1,1248 @@
+import pytest
```
These files should be removed now that they're under `test/units/`.

mattclay commented on test/units/module_utils/network/aci/aci_construct_url_test.py:
```diff
@@ -0,0 +1,1248 @@
+import pytest
```
Test files under `test/units/` need to be named `test_{something}.py` instead of `{something}_test.py`.

RobW3LGA committed:
Deleting for file name change per Matt Clay

RobW3LGA committed:
Deleting for file name change per Matt Clay

RobW3LGA committed:
Correcting file names per Matt Clay

@mattclay Thanks again for your continued guidance and patience. Please cancel the previous (incorrect) request

RobW3LGA committed:
Wrong location for test file

RobW3LGA committed:
Wrong location for test file

RobW3LGA committed:
First attempt to comply with suggestions

lib/ansible/module_utils/network/aci/aci.py:517:0: SyntaxWarning: "is not" with a literal. Did you mean "!="?
lib/ansible/module_utils/network/aci/aci.py:534:0: SyntaxWarning: "is not" with a literal. Did you mean "!="?
lib/ansible/module_utils/network/aci/aci.py:558:161: E501 line too long (210 > 160 characters)

RobW3LGA committed:
First attempt to comply with suggestions

test/units/module_utils/network/aci/test_aci_construct_url.py:1:14: SyntaxError: import pytest
test/units/module_utils/network/aci/test_aci_deep_url.py:1:14: SyntaxError: import pytest
test/units/module_utils/network/aci/test_aci_construct_url.py:0:0: use "\n" for line endings instead of "\r\n"
test/units/module_utils/network/aci/test_aci_deep_url.py:0:0: use "\n" for line endings instead of "\r\n"
Shortened test function names (less descriptive)

RobW3LGA committed:
Second attempt to comply with suggestions

RobW3LGA committed:
Second attempt to comply with suggestions

RobW3LGA committed:
Third attempt to comply with suggestions

RobW3LGA committed:
Third attempt to comply with suggestions

RobW3LGA committed:
Pro Tip: Convert from 'CRLF' to 'LF' in VSCode

It's on the status bar to the right

RobW3LGA committed:
Added setup() support for tests

RobW3LGA committed:
Continued corrections to support testing

RobW3LGA committed:
Added two mocks to support testing

I could not find where to place fakes/mocks, so please let me know if the current location is incorrect

RobW3LGA committed:
Adding tmpdir property to mock_basic.py

RobW3LGA committed:
Added last blank line to mock_basic.py

To pass sanity test

RobW3LGA committed:
Attempt to correct setup() issues

RobW3LGA committed:
Attempt to correct setup() issues

RobW3LGA committed:
Attempt to correct setup() issues

RobW3LGA committed:
Attempt to correct setup() issues

RobW3LGA committed:
Withdrawing pending injectability tweak to aci.py

RobW3LGA committed:
Withdrawing pending injectability tweak to aci.py

RobW3LGA committed:
Withdrawing pending injectability tweak to aci.py

RobW3LGA committed:
Withdrawing pending injectability tweak to aci.py

RobW3LGA committed:
Merge pull request #1 from ansible/devel

Re-sync forked repo

dagwieers commented on lib/ansible/module_utils/network/aci/aci.py:
```diff
@@ -446,7 +630,7 @@ def construct_url(self, root_class, subclass_1=None, subclass_2=None, subclass_3
         if self.child_classes:
             # Append child_classes to filter_string if filter string is empty
-            self.update_qs({'rsp-subtree': 'full', 'rsp-subtree-class': ','.join(self.child_classes)})
+            self.update_qs({'rsp-subtree': 'full', 'rsp-subtree-class': ','.join(sorted(self.child_classes))})
```
Why sorted?

RobW3LGA replied:
When under test, self.child_classes appears to randomize the response when .join() is called (the unit test randomly fails). Sorted() eliminates the randomization and allows the test to pass consistently. My next PR will allow aci.py to be testable and I will be able to demonstrate this with the unit tests I have ready to upload.

RobW3LGA commented:
To any of the reviewers: I'm not sure how to proceed with this PR. Am I missing something you all need to review or proceed with? Once this PR is approved I do have quite a few useful contributions waiting (like a process that generates modules quickly and consistently).

dagwieers commented:
@RobW3LGA I didn't want to add this as part of the v2.8 release as it came late to the game. But the approach is the right one, and we need to convert all the modules over.
